src/transformers/modeling_tf_bart.py

```diff
@@ -0,0 +1,1215 @@
+
+        if attn_mask is not None:
+            assert attn_mask.dtype == tf.float32, f"expected dtype tf.float32 got {attn_mask.dtype}"
+            attn_weights = tf.reshape(attn_weights, (bsz, self.num_heads, tgt_len, src_len)) + attn_mask
+            attn_weights = tf.reshape(attn_weights, (bsz * self.num_heads, tgt_len, src_len))
+
+        if key_padding_mask is not None and key_padding_mask._rank() == 0:
+            key_padding_mask = None
+        assert key_padding_mask is None or key_padding_mask.shape[:2] == (
+            bsz,
+            src_len,
+        )
+        if key_padding_mask is not None:  # don't attend to padding symbols
+            attn_weights: tf.Tensor = tf.reshape(attn_weights, (bsz, self.num_heads, tgt_len, src_len))
+            if key_padding_mask.dtype == tf.bool:
+                key_padding_mask = tf.cast(key_padding_mask, attn_weights.dtype) * -1e9
+            extended_mask = tf.expand_dims(tf.expand_dims(key_padding_mask, 1), 2)
+            attn_weights = attn_weights + extended_mask
+            attn_weights = tf.reshape(attn_weights, (bsz * self.num_heads, tgt_len, src_len))
+
+        attn_weights = tf.nn.softmax(attn_weights, axis=-1)
+        attn_probs = tf.nn.dropout(attn_weights, rate=self.dropout if training else 0.0)
+
+        assert v is not None
+        attn_output = tf.matmul(attn_probs, v)
+        assert attn_output.shape == (bsz * self.num_heads, tgt_len, self.head_dim)
+        attn_output = tf.transpose(attn_output, perm=(1, 0, 2))
+        attn_output = tf.reshape(attn_output, (tgt_len, bsz, embed_dim))
+        attn_output = self.out_proj(attn_output)
+        attn_weights: tf.Tensor = tf.reshape(attn_weights, (bsz, self.num_heads, tgt_len, src_len))
+        return attn_output, attn_weights
+
+    def _use_saved_state(self, k, v, saved_state, key_padding_mask, static_kv, bsz):
+        # saved states are stored with shape (bsz, num_heads, seq_len, head_dim)
+        if "prev_key" in saved_state:
+            _prev_key = saved_state["prev_key"]
+            assert _prev_key is not None
+            prev_key = tf.reshape(_prev_key, (bsz * self.num_heads, -1, self.head_dim))
+            if static_kv:
+                k = prev_key
+            else:
+                assert k is not None
+                k = tf.concat([prev_key, k], axis=1)
+        if "prev_value" in saved_state:
+            _prev_value = saved_state["prev_value"]
+            assert _prev_value is not None
+            prev_value = tf.reshape(_prev_value, (bsz * self.num_heads, -1, self.head_dim))
+            if static_kv:
+                v = prev_value
+            else:
+                assert v is not None
+                v = tf.concat([prev_value, v], axis=1)
+        assert k is not None and v is not None
+        prev_key_padding_mask = saved_state.get("prev_key_padding_mask", None)  # type: Optional[Tensor]
+        key_padding_mask = self._cat_prev_key_padding_mask(
```

Comment:
I don't really get why we need all this logic for the attention mask. The attention_mask is automatically increased when doing `generate()` and when using `use_cache=True` we should use the same `attention_mask` IMO. I highly doubt that we need all this complex logic for cached generation TBH.

Reply:
Addressed!